airbyte-integrations/connectors/source-microsoft-onedrive/source_microsoft_onedrive/stream_reader.py

```diff
@@ -0,0 +1,178 @@
+
+    def get_files_by_drive_name(self, drives, drive_name, folder_path):
+        """Yields files from the specified drive."""
+        path_levels = [level for level in folder_path.split("/") if level]
+        folder_path = "/".join(path_levels)
+
+        for drive in drives:
+            is_onedrive = drive.drive_type in ["personal", "business"]
+            if drive.name == drive_name and is_onedrive:
+                folder = drive.root.get_by_path(folder_path).get().execute_query()
+                yield from self.list_directories_and_files(folder)
+
+    def get_matching_files(self, globs: List[str], prefix: Optional[str], logger: logging.Logger) -> Iterable[RemoteFile]:
+        """
+        Retrieve all files matching the specified glob patterns in OneDrive.
+        """
+        drives = self.one_drive_client.drives.get().execute_query()
+
+        if self.config.credentials.auth_type == "Client":
+            my_drive = self.one_drive_client.me.drive.get().execute_query()
+        else:
+            my_drive = (
+                self.one_drive_client.users.get_by_principal_name(self.config.credentials.user_principal_name).drive.get().execute_query()
+            )
+
+        drives.add_child(my_drive)
+
+        files = self.get_files_by_drive_name(drives, self.config.drive_name, self.config.folder_path)
+
+        try:
+            first_file = next(files)
+            yield from self.filter_files_by_globs_and_start_date(
+                [
+                    MicrosoftOneDriveRemoteFile(
+                        uri=first_file.name,
+                        download_url=first_file.properties["@microsoft.graph.downloadUrl"],
+                        last_modified=first_file.properties["lastModifiedDateTime"],
+                    )
+                ],
+                globs,
+            )
+
+        except StopIteration as e:
+            raise AirbyteTracedException(
+                internal_message=str(e),
+                message=f"Drive '{self.config.drive_name}' is empty or does not exist.",
+                failure_type=FailureType.config_error,
+                exception=e,
+            )
+
+        yield from self.filter_files_by_globs_and_start_date(
+            [
+                MicrosoftOneDriveRemoteFile(
+                    uri=file.name,
+                    download_url=file.properties["@microsoft.graph.downloadUrl"],
+                    last_modified=file.properties["lastModifiedDateTime"].strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
+                )
+                for file in files
+            ],
+            globs,
+        )
+
+    @contextmanager
+    def open_file(self, file: RemoteFile, mode: FileReadMode, encoding: Optional[str], logger: logging.Logger) -> IOBase:
+        """
+        Context manager to open and yield a remote file from OneDrive for reading.
+        """
+        try:
+            with smart_open.open(file.download_url, mode=mode.value, encoding=encoding) as file_handle:
```

Comment:
For file type unstructured I'm getting this error message on check:
> {"type": "CONNECTION_STATUS", "connectionStatus": {"status": "FAILED", "message": "['Unable to connect to stream <airbyte_cdk.sources.file_based.stream.default_file_based_stream.DefaultFileBasedStream object at 0x1236a8e80> - Traceback (most recent call last):\\n  File \"/Users/joereuter/Clones/airbyte/airbyte-integrations/connectors/source-microsoft-onedrive/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/file_based/file_based_source.py\", line 80, in check_connection\\n    ) = stream.availability_strategy.check_availability_and_parsability(stream, logger, self)\\n  File \"/Users/joereuter/Clones/airbyte/airbyte-integrations/connectors/source-microsoft-onedrive/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/file_based/availability_strategy/default_file_based_availability_strategy.py\", line 64, in check_availability_and_parsability\\n    handle.close()\\nAttributeError: \\'_GeneratorContextManager\\' object has no attribute \\'close\\'\\n']"}}

It seems like this can be simplified by just returning the IOBase, as it's a context manager itself:
```
def open_file(self, file: RemoteFile, mode: FileReadMode, encoding: Optional[str], logger: logging.Logger) -> IOBase:
        """
        Context manager to open and yield a remote file from OneDrive for reading.
        """
        try:
            return smart_open.open(file.download_url, mode=mode.value, encoding=encoding)
        except Exception as e:
            logger.exception(f"Error opening file {file.uri}: {e}")
```

See the same in s3: https://github.com/airbytehq/airbyte/blob/a41d11b4a024c9715cf4387802bc88c5c918453c/airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py#L121

Reply:
fixed